File: packages/go/slop-ai/unix_client.go
package slop

import (
	"bufio"
	"context"
	"encoding/json"
	"net"
	"sync"
)

// UnixClientTransport connects to a SLOP provider over a Unix domain socket using NDJSON.
type UnixClientTransport struct {
	Path string
}

type unixClientConn struct {
	conn         net.Conn
	mu           sync.Mutex
	msgHandler   func(map[string]any)
	closeHandler func()
}

// Connect dials the Unix socket and returns a ClientConnection.
func (t *UnixClientTransport) Connect(_ context.Context) (ClientConnection, error) {
	conn, err := net.Dial("unix", t.Path)
	if err != nil {
		return nil, err
	}
	uc := &unixClientConn{conn: conn}
	go uc.readLoop()
	return uc, nil
}

func (uc *unixClientConn) Send(msg map[string]any) error {
	data, err := json.Marshal(msg)
	if err != nil {
		return err
	}
	data = append(data, '\n')
	uc.mu.Lock()
	defer uc.mu.Unlock()
	_, err = uc.conn.Write(data)
	return err
}

func (uc *unixClientConn) OnMessage(handler func(map[string]any)) {
	uc.mu.Lock()
	defer uc.mu.Unlock()
	uc.msgHandler = handler
}

func (uc *unixClientConn) OnClose(handler func()) {
	uc.mu.Lock()
	defer uc.mu.Unlock()
	uc.closeHandler = handler
}

func (uc *unixClientConn) Close() error {
	return uc.conn.Close()
}

func (uc *unixClientConn) readLoop() {
	defer func() {
		uc.mu.Lock()
		handler := uc.closeHandler
		uc.mu.Unlock()
		if handler != nil {
			handler()
		}
	}()

	scanner := bufio.NewScanner(uc.conn)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}
		var msg map[string]any
		if json.Unmarshal([]byte(line), &msg) != nil {
			continue
		}
		uc.mu.Lock()
		handler := uc.msgHandler
		uc.mu.Unlock()
		if handler != nil {
			handler(msg)
		}
	}
}

File: packages/python/slop-ai/src/slop/transports/unix_client.py
"""Unix domain socket client transport for SlopConsumer."""

from __future__ import annotations

import asyncio
import json
from typing import Any, Callable


class UnixClientConnection:
    """Client connection over Unix domain socket (NDJSON framing)."""

    def __init__(
        self,
        reader: asyncio.StreamReader,
        writer: asyncio.StreamWriter,
    ) -> None:
        self._reader = reader
        self._writer = writer
        self._message_handler: Callable[[dict[str, Any]], None] | None = None
        self._close_handler: Callable[[], None] | None = None
        self._reader_task: asyncio.Task[None] | None = None

    async def send(self, message: dict[str, Any]) -> None:
        data = json.dumps(message) + "\n"
        self._writer.write(data.encode())
        await self._writer.drain()

    def on_message(self, handler: Callable[[dict[str, Any]], None]) -> None:
        self._message_handler = handler

    def on_close(self, handler: Callable[[], None]) -> None:
        self._close_handler = handler

    async def close(self) -> None:
        if self._reader_task is not None:
            self._reader_task.cancel()
        self._writer.close()
        await self._writer.wait_closed()

    def _start_reader(self) -> None:
        self._reader_task = asyncio.get_running_loop().create_task(self._read_loop())

    async def _read_loop(self) -> None:
        try:
            while True:
                line = await self._reader.readline()
                if not line:
                    break
                if self._message_handler is not None:
                    msg: dict[str, Any] = json.loads(line)
                    self._message_handler(msg)
        except asyncio.CancelledError:
            return
        except Exception:
            pass
        finally:
            if self._close_handler is not None:
                self._close_handler()


class UnixClientTransport:
    """Client transport that connects to a Unix domain socket."""

    def __init__(self, path: str) -> None:
        self._path = path

    async def connect(self) -> UnixClientConnection:
        reader, writer = await asyncio.open_unix_connection(self._path)
        conn = UnixClientConnection(reader, writer)
        conn._start_reader()
        return conn
